packages/aragon-api/src/index.js

```diff
@@ -228,19 +308,74 @@ export class AppProxy {
       )
 
-    const store$ = initialState.pipe(
-      switchMap((initialState) =>
-        merge(
-          this.events(),
-          ...events
-        ).pipe(
-          mergeScan(wrappedReducer, initialState, 1),
-          map((state) => this.cache('state', state))
+    const getCurrentEvents = (fromBlock) => merge(
+      this.events(fromBlock),
+      ...externals.map(({ contract }) => contract.events(fromBlock))
+    )
+
+    // If `cachedFromBlock` is null there's no cache, `pastEvents` will use the initialisationBlock
+    // External contracts can specify their own `initializationBlock` which will be used in case the cache is empty
+    const getPastEvents = (cachedFromBlock, toBlock) => merge(
+      this.pastEvents(cachedFromBlock, toBlock),
+      ...externals.map(({ contract, initializationBlock }) => contract.pastEvents(cachedFromBlock || initializationBlock, toBlock))
+    )
+
+    const cachedState$ = this.state().pipe(first())
+    const cachedBlock$ = this.getCache(CACHED_BLOCK_KEY)
+    const latestBlock$ = this.web3Eth('getBlockNumber')
+    const initState$ = init ? from(init()) : from([null])
+
+    const store$ = combineLatest(cachedState$, initState$, cachedBlock$, latestBlock$).pipe(
+      switchMap(([cachedState, initState, cachedBlock, latestBlock]) => {
+        console.debug('- store - initState', initState)
+        console.debug('- store - cachedState', cachedState)
+        console.debug(`- store - cachedBlock ${cachedBlock} | latestBlock: ${latestBlock}`)
+
+        // The block up to which to fetch past events.
+        // The reduced state up to this point will be cached on every load
+        const pastEventsToBlock = latestBlock - BLOCK_REORG_MARGIN
+
+        console.debug(`- store - pastEvents: ${cachedBlock} -> ${pastEventsToBlock} (${pastEventsToBlock - cachedBlock} blocks)`)
+        console.debug(`- store - currentEvents$: from: ${pastEventsToBlock} -> future`)
+        // currentEvents$ is an observable for events starting from pastEventsToBlock
+        const currentEvents$ = getCurrentEvents(pastEventsToBlock)
+
+        const pastState$ = getPastEvents(cachedBlock, pastEventsToBlock).pipe(
+          // Prevent multiple subscriptions invoking duplicate calls
+          share(),
+          // single emission array of all pastEvents -> flatten to process events
+          flatMap(pastEvents => from(pastEvents)),
+          mergeScan(wrappedReducer, { ...cachedState, ...initState }, 1),
+          // debounce to reduce rendering and caching overthead
+          debounceTime(200),
+          tap((state) => {
+            this.cache(CACHED_BLOCK_KEY, pastEventsToBlock)
+            this.cache('state', state)
+          })
         )
-      ),
+
+        const currentState$ = pastState$.pipe(
+          // Use the last past state as the initial state for reducing current/future states
+          last(),
+          switchMap(pastState => {
+            return merge(currentEvents$, accounts$).pipe(
```

Comment:
> Can we define currentEvents in this block?

Good idea.

> ⚠️ Also, I think we may have a problem with the accounts$ observable being hot; ideally we would only kick it off here (the old semi-hot behaviour will not be a problem anymore after #305).

We only really care about the last value which `publishReplay` should ensure is available. Do you think it fits better here even if we don't have a hot observable problem? 
I put it in the top block to avoid bloating this part with all the object creation part.

Reply:
Ah yes, that's true! I think with #305 we will be able to just use `app.accounts()` to simplify this whole bit.